File: src/main/java/com/vc/easy/L844.scala
package com.vc.easy

object L844 {
  def backspaceCompare(S: String, T: String): Boolean = {
    import scala.collection.mutable
    def solve(str: String): String = {
      val st = new mutable.Stack[Char]()
      str.foreach(ch => {
        if(ch == '#') {
          if(st.nonEmpty) st.pop
        }
        else st.push(ch)
      })
      st.mkString("")
    }
    solve(S) == solve(T)
  }
}

File: src/main/java/com/vc/easy/L844.java
package com.vc.easy;

import java.util.Stack;

class L844 {
    public boolean backspaceCompare(String S, String T) {
        Stack<Character> sts = new Stack<>();
        for(int i = 0; i < S.length(); i++) {
            if(S.charAt(i) == '#') {
                if(!sts.isEmpty()) sts.pop();
            }
            else sts.push(S.charAt(i));
        }

        Stack<Character> stt = new Stack<>();
        for(int i = 0; i < T.length(); i++) {
            if(T.charAt(i) == '#') {
                if(!stt.isEmpty()) stt.pop();
            }
            else stt.push(T.charAt(i));
        }

        while(!sts.isEmpty() && !stt.isEmpty()) {
            if(sts.pop() != stt.pop()) return false;
        }

        return sts.isEmpty() && stt.isEmpty();
    }
}
